### ISSUE TITLE
pgr_createtopology() does not recognize temporary tables.

### ISSUE THREAD

**Issue by Rotiahn** [pgRouting/pgrouting#252]
*TL;DR: I think that a recent change (between 2.0.0-rc1 and 2.0.0) is preventing pgr_createtopology() from operating on temporary tables (and possibly other functions as well).  I think the problem comes from a change in pgr_getTableName().  I found a change to make the function work (but don't know what side effects it would have).*

**My environment**
Debian 6.0.7
PostgreSQL version 9.2.4 (compiled from source)
PostGIS version 2.0.3 (compiled from source)
Upgraded pgrouting from 2.0.0-rc1 to 2.0.0 (compiled from source)

**pgrouting-v2.0.0-rc1:**
>=# select pgr_createtopology('temporary table', 0.001, 'geo', 'row_number'); 
> *results in successful creation of topology*

**pgrouting-v2.0.0:**
>=# select pgr_createtopology('temporary table', 0.001, 'geo', 'row_number'); 
>NOTICE:  PROCESSING:
>NOTICE: pgr_createTopology('xo_fiber_numbered',0.001,'geo','row_number','source','target','>true')
>NOTICE:  Performing checks, pelase wait .....
>NOTICE:  -------> xo_fiber_numbered not found
>pgr_createtopology
> \--------------------
> FAIL
>(1 row)

>select pgr_createtopology('**regular table**', 0.001, 'geo', 'row_number'); 
> *results in successful creation of topology*

Based on issue from osm2pgsql:https://github.com/pgRouting/osm2pgrouting/issues/28  I suspect the problem comes from: pgrouting_utilities.sql. 

Removing the requirement of table_type='BASE TABLE' from function pgr_gettablename() appears to allow the below version to work (temporary tables have a table_type='LOCAL TEMPORARY':

> =# select pgr_createtopology('pg_temp_2.temporary table', 0.001, 'geo', 'row_number'); 

**Subsection of changed pgr_getTableName():**
>    if sname is NOT NULL THEN -- found schema (as is)
>        EXECUTE 'select table_name from information_schema.tables where ' 
>--        ||'table_type='||quote_literal('BASE TABLE')||' and '
>        ||'table_schema='||quote_literal(sname)||' and '
>        ||'table_name='||quote_literal(tn) INTO  naming;

I can post my full changed function if that would help.

**Comment by cvvergara** [comment]
CreateTopology has this well documented in big letters warning:
http://docs.pgrouting.org/2.0/en/src/common/doc/functions/create_topology.html#pgr-create-topology
maybe the documentation lacks that it doenst work for temporary tables because:
Might happen that to a temporary table can not create temporary geometry.  (and I am not catching them to give back an appropiate message)
It is also creating a vertices table that is not temporary based on the name so its easy to relate which vertex table is related to the edge table,  we create a temporary vertex table table during the process.
By the way, pgrouting_topology.sql#L270 is the only thing to be changed because it accepts bigint original columns. Can be modifyed to: if the user's types are bigint then create the table with bigint, if the users types are int generate with int.

**Comment by woodbri** [comment]
As the bug documents it is likely that this issues is being caused by pgr_getTableName() not functioning correctly for used with temporary tables. And therefore all code depending on pgr_getTableName() will also likely fail.

```
create temporary table temp_abc (
  id integer not null primary key,
  junk text
);

create table nottemp_abc (
  id integer not null primary key,
  junk text
);

select * from pgr_getTableName('nottemp_abc');  -- this works
select * from pgr_getTableName('temp_abc');  -- this fails to return the table
```

**Comment by cvvergara** [comment]
After inspecting the code  and checked that BASE TABLE is the only type accepted, I have this questions:
What table types do you want it to accept it accept? are they also bugs? are they features?
VIEW for a view, FOREIGN TABLE for a foreign table, or LOCAL TEMPORARY for a temporary table.
does the index creation works for a view or for a foreign table (whatever that is).?
What are the side effects of createTopology does to the database tables?
Can we modify a view automatically?
A temporary table is it going to create a temporary vertex table?
The actual naming conventions that I made because it was always writing over the same table.
Is it going to be writing over the same table because its a temporary?  we going to change that naming convention? 
if not, then what is going to be the new naming conventions? and changing the naming convention  is going to affect currently client-server applications?
Those are questions that have to be answered so that we can clearly document the table types createTopology uses. 
but for standarization.
Other functions like nodeNetwork/analize topology have this so called bug, so what is the action to take?
which ones?, do they have the same restrictions like they don't affect the database?
Can they have the same solution?

**Comment by woodbri** [comment]
Ok, great questions. I'll do my best to address them.

* on one hand we can simplify the whole mess by only supporting base tables and throwing an appropriate error for other tables and documenting this restriction.

IF (and I'm not saying we should), here are my thoughts on the other table types.

* FOREIGN TABLE, I have never used and in fact had to lookup what it was. So, I'm probably no qualified to comment, but that never stopped me. While the idea is interesting, I'm not sure how popular these are. I wonder if postgis supports them? Can you update a remote foreign table? I would probably take a pass on this one for now, but if it behaves just like a base table without any restriction that would impact pgRouting from using them, I'm fine with adding support for it. The vertex table would be a local base table to keep things simple.

* VIEW, views could be handy, but they also require rules and triggers defined in them so you can update the records. Under my "keep it simple" rule, I'm inclined to take a pass on this one. (see extra table note below)

* TEMPORARY, I'm not sure why people are using temporary tables except for potentially testing purposes because these are dropped when the connection to the database is closed. While the vertices table could be also be temporary and hence automatically deleted, I like to keep them around because I find them useful for analyzing the network and for plotting information in them using mapserver. I suppose if the user passes a temporary table as input, that we could create a vertices table also as temporary.

One thing that might be a useful enhancement is to add a flag to indicate it we want the original table updated with source and target columns or if we should create a new table like:

```
create table edge_topo (
   integer id not null primary key,  -- edge table id field
   integer source, -- source id for this edge
   integer target -- target id for this edge
);
```
then this can be joined to the edge table in the query passed to pgr_* function.

So bottom line: There is stuff we could do to better support these various table types, but I think for 2.1.0 we should probably only support base tables and fix the code to throw an appropriate error. And then for 2.2.0+, think about adding support for some or none of the above features. I could be swayed in another direction if anyone has strong thoughts on this? Change it to Milestone 2.2.0??

**Comment by cvvergara** [comment]
Let me put this two scenarios: (maybe it can help on what actions has to be done)

1) When routing data comes from osm, using osm2pgrouting or osm2po, the information comes already populated with the topology, so these users don't need to create a topology.

2) Personally, I don't use osm data, so I had to use the function to create the topology, in this case the topology of a whole country, process that took quite a while: I can't really remember how much it took, but it was in the order of over 20 hours, but it was a one time operation, the topology of the whole country is there; if I am working with a particular bounding box, I get the topology of the segments that are in the bounding box.

**Comment by woodbri** [comment]
I am inclined to close this as wont fix because I do not see a use case for it other than potentially testing. Ans there is a simple work alternative to the problem by using base tables rather than temporary tables and if the user want them deleted when done then to drop the tables.

**Comment by Rotiahn** [comment]
For What it's worth, I figured I'd share the use-case that triggered me to report this error originally:

Note:  I haven't touched the below project or client for quite a while, so they may not necessarily still be doing it this way.  Nor am I claiming it is the best-approach for solving this problem, just an example:

Source Data is Client's proprietary map of GIS network segments.  It's composed purely of a long list of line-strings of 2+ vertices.  They only track nodes if actual physical infrastructure intersects (i.e. just because two line-strings intersect != node).  On demand, they want to pick two points and get the full list of possible routes between those points based on their topology.  

I was using temporary tables to build the appropriate vertex and route tables because after the analysis was done, the tables needed to be removed and rebuilt for the next analysis (to ensure that any changes to the network were incorporated, ensure that parallel analyses were kept isolated, and ensure that space was freed between runs).  There may have been other reasons for using temporary tables, but these are the ones I can remember.

Using base tables probably would have sufficed, but would have meant being more careful to delete my tables afterward.  And it might have run the risk of cross-contamination with parallel instances of the analysis if I wasn't careful.

Removing the requirement that tables be Base tables appeared to solve the the problem with my existing code without causing any unpleasant side effects for at least as long as the project lasted (it was cancelled before production).  So I did not explore further.

**Comment by cvvergara** [comment]
I really don't think this can be done.

**Comment by iboates** [comment]
I realize that this discussion has already been had and a decision was taken, but is there any chance of it being added in version 3?
